lib/delta.ex

```diff
@@ -590,28 +590,39 @@ defmodule Delta do
     other = push(other_rest, other_remaining)
 
     delta =
-      case {base_op, other_op} do
-        {%{"insert" => ins}, %{"insert" => ins}} ->
-          attrs = Delta.Attr.diff(base_op["attributes"], other_op["attributes"])
-          push(delta, Op.retain(op_len, attrs))
-
+      {base_op, other_op}
+      |> case do
         {%{"insert" => base_insert}, %{"insert" => other_insert}}
         when is_map(base_insert) and is_map(other_insert) ->
-          {embed_type, base_embed, other_embed} = Op.get_embed_data!(base_insert, other_insert)
-
-          case get_handler(embed_type) do
-            nil ->
-              delta
-              |> push(Op.delete(op_len))
-              |> push(other_op)
-
-            handler ->
-              diff = handler.diff(base_embed, other_embed)
-              attrs = Delta.Attr.diff(base_op["attributes"], other_op["attributes"])
-              push(delta, Op.retain(%{embed_type => diff}, attrs))
-          end
+          diff_embeds(base_op, other_op)
+
+        _ ->
+          diff_simple_ops(base_op, other_op)
       end
+      |> Enum.reduce(delta, fn op, acc -> push(acc, op) end)
 
     do_diff(base, other, diffs, delta, :equal, len - op_len)
   end
+
+  defp diff_embeds(base_op = %{"insert" => base_insert}, other_op = %{"insert" => other_insert}) do
+    with [{base_embed_type, _embed}] <- Map.to_list(base_insert),
+         [{other_embed_type, _embed}] <- Map.to_list(other_insert),
+         true <- base_embed_type == other_embed_type,
```

Comment:
I think you could avoid checking if the two embed types are equal by pinning the embed type when matching the second `Map.to_list`:

```elixir
with [{embed_type, _embed}] <- Map.to_list(base_insert),
     [{^embed_type, _embed}] <- Map.to_list(other_insert),
```

Reply:
Neat! Didn't think of that